modules/tinymce/src/core/main/ts/keyboard/BoundarySelection.ts

```diff
@@ -35,9 +36,19 @@ const renderCaretLocation = (editor: Editor, caret: Cell<Text>, location: Bounda
     return location;
   });
 
+const getPositionFromRange = (range: Range, root: HTMLElement, forward: boolean): CaretPosition => {
+  const start = CaretPosition.fromRangeStart(range);
+  if (range.collapsed) {
+    return start;
+  } else {
+    const end = CaretPosition.fromRangeEnd(range);
+    return forward ? CaretFinder.prevPosition(root, end).getOr(end) : CaretFinder.nextPosition(root, start).getOr(start);
```

Comment:
I haven't looked into this much yes, but is this correct? I would have expected the start position to be the same regardless of if it was collapsed. This instead will make it move back inside the inline boundary element.

Reply:
Without rebasing position to previous one  caret can not leave inline boundary element

https://user-images.githubusercontent.com/44773772/165530437-afba2b99-4d8f-43ca-ba3d-0ae7f7107e1a.mov

Reply:
I tested this locally and the current behaviour with these changes isn't correct. With these changes it causes it to effectively just collapse the selection, but it shouldn't be doing that it should be moving to the next viable caret location. So lets say I have:
```html
<p>Some <a href="#">content</a> here.</p>
```
So in the above, lets say we select the inner content of the anchor. For example (where `|` is the selection):
```html
<p>Some <a href="#">|content|</a> here.</p>
```
Now, when I press the left key the selection should be in the before the anchor element for example:
```html
<p>Some |<a href="#">content</a> here.</p>
```
Instead all it's doing is collapsing it so it's still inside the anchor at the very first caret position.

Likewise in reverse, if we pressed the right arrow key instead then I'd expect it to end up here:
```html
<p>Some <a href="#">content</a>| here.</p>
```

I haven't had a chance to look into this more, but something needs to change here to fix the above. We also should add tests for cases like the above, as the current ones only handle if the outside edges of the anchor are selected but not the inside edges. So it'd need to be something like this:
```
editor.setContent('<p>test <span class="mce-annotation">span</span> selection</p>', { format: 'raw' });
TinySelections.setSelection(editor, [ 0, 1, 0 ], 0, [ 0, 1, 0 ], 4, true);
TinyContentActions.keystroke(editor, Keys.left());
TinyAssertions.assertCursor(editor, [ 0, 0 ], 5);
```

- [ ] Add tests for the cases above and fix issues

Reply:
Actually, really sorry @vpyshnenko. In my haste I realised I've got this backwards and your current behaviour is correct for the case mentioned above (i.e. it should collapse, which is what the browser does natively as well).

Anyways, it'd probably still be good to add some tests for where the selection is inside and then we press left/right if you don't mind?